hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java

```diff
@@ -837,10 +840,75 @@ public void update(HoodieCleanMetadata cleanMetadata, String instantTime) {
    */
   @Override
   public void update(HoodieRestoreMetadata restoreMetadata, String instantTime) {
-    processAndCommit(instantTime, () -> HoodieTableMetadataUtil.convertMetadataToRecords(engineContext,
-        metadataMetaClient.getActiveTimeline(), restoreMetadata, getRecordsGenerationParams(), instantTime,
-        metadata.getSyncedInstantTime()), false);
-    closeInternal();
+    dataMetaClient.reloadActiveTimeline();
+
+    // Since the restore has completed on the dataset, the latest write timeline instant is the one to which the
+    // restore was performed. This should be always present.
+    final String restoreToInstantTime = dataMetaClient.getActiveTimeline().getWriteTimeline()
+        .getReverseOrderedInstants().findFirst().get().getTimestamp();
+
+    // We cannot restore to before the oldest compaction on MDT as we don't have the basefiles before that time.
+    Option<HoodieInstant> lastCompaction = metadataMetaClient.getCommitTimeline().filterCompletedInstants().lastInstant();
+    if (lastCompaction.isPresent()) {
+      if (HoodieTimeline.LESSER_THAN_OR_EQUALS.test(restoreToInstantTime, lastCompaction.get().getTimestamp())) {
+        String msg = String.format("Cannot restore MDT to %s because it is older than latest compaction at %s", restoreToInstantTime,
+            lastCompaction.get().getTimestamp()) + ". Please delete MDT and restore again";
+        LOG.error(msg);
+        throw new HoodieMetadataException(msg);
+      }
+    }
+
+    // Restore requires the existing pipelines to be shutdown. So we can safely scan the dataset to find the current
+    // list of files in the filesystem.
+    List<DirectoryInfo> dirInfoList = listAllPartitions(dataMetaClient);
+    Map<String, DirectoryInfo> dirInfoMap = dirInfoList.stream().collect(Collectors.toMap(DirectoryInfo::getRelativePath, Function.identity()));
+    dirInfoList.clear();
+
+    LOG.info("Restoring MDT to " + restoreToInstantTime + " at " + instantTime);
+    getWriteClient().restoreToInstant(restoreToInstantTime, false);
+
+    // At this point we have also reverted the cleans which have occurred after the restoreToInstantTime. Hence, a sync
```

Comment:
do we have tests covering this scenario?

Reply:
TestHoodieBackedMetadata#testTableOperationsWithRestore

Reply:
I mean, specifically where we have cleans in the timeline which is being restored. I did happen to check the test testTableOperationsWithRestore, but I don't think it tests this case

Reply:
The impli is kind of ugly here, we want to rollback to a specific history time point of the MDT, but we also want to keep the commits for cleaning, is there any possibility we only rollback the commit(we don't want) instead of perform a `restoreToInstant` ?

Reply:
Ignore me, this is a restore operation itself.

Reply:
TestHoodieBackedMetadata#testTableOperationsWithRestore

Clean is performed with timestamp 0000009
After the clean, we restore to 0000006 which is older than clean timestamp.

So the above test should test this scenario.